consensus/istanbul/core/commit.go

```diff
@@ -207,62 +218,135 @@ func (c *core) handleCheckedCommitForCurrentSequence(msg *istanbul.Message, comm
+				func(c *istanbul.CommittedSubject) []byte { return c.EpochValidatorSetSeal },
+			)
+			if err != nil {
+				nextRound := new(big.Int).Add(c.current.Round(), common.Big1)
+				logger.Warn("Error on commit, waiting for desired round", "reason", "getAggregatedSeal", "err", err, "desired_round", nextRound)
+				c.waitForDesiredRound(nextRound)
+				return nil
+			}
+
+			aggregatedEpochValidatorSetSeal.Bitmap = epochBitmap
+			aggregatedEpochValidatorSetSeal.Signature = epochAggregate
 		}
 
-	} else if c.current.GetPrepareOrCommitSize() >= minQuorumSize && c.current.State().Cmp(StatePrepared) < 0 {
-		err := c.current.TransitionToPrepared(minQuorumSize)
+		logger.Trace("Got a quorum of commits", "tag", "stateTransition", "commits", commits)
+		err = c.commit(aggregatedSeal, aggregatedEpochValidatorSetSeal)
 		if err != nil {
-			logger.Error("Failed to create and set prepared certificate", "err", err)
+			logger.Error("Failed to commit()", "err", err)
 			return err
 		}
-		// Process Backlog Messages
-		c.backlog.updateState(c.current.View(), c.current.State())
-
-		logger.Trace("Got quorum prepares or commits", "tag", "stateTransition", "commits", c.current.Commits, "prepares", c.current.Prepares)
-		c.sendCommit()
 	}
 	return nil
 
 }
 
+// generateAggregateCommittedSeal will generate the aggregate committed seal
+// verify it and return it.
+func (c *core) generateAggregateCommittedSeal() (types.IstanbulAggregatedSeal, error) {
+	commits := c.current.Commits()
+	if commits.Size() < c.current.ValidatorSet().MinQuorumSize() {
+		return types.IstanbulAggregatedSeal{}, fmt.Errorf("insufficient commits to construct aggregate seal")
+	}
+	bitmap, aggregate, err := AggregateSeals(
+		commits,
+		func(c *istanbul.CommittedSubject) []byte { return c.CommittedSeal },
+	)
+	if err != nil {
+		return types.IstanbulAggregatedSeal{}, fmt.Errorf("failed to aggregate seals: %v", err)
+	}
+
+	aggregatedSeal := types.IstanbulAggregatedSeal{Bitmap: bitmap, Signature: aggregate, Round: c.current.Round()}
+	err = c.backend.VerifyAggregatedSeal(c.current.Proposal().Hash(), c.current.ValidatorSet(), aggregatedSeal)
+	if err != nil {
+		return types.IstanbulAggregatedSeal{}, fmt.Errorf("failed verify aggregate seal: %v", err)
+	}
+	return aggregatedSeal, nil
+}
+
+// removeInvalidCommittedSeals individually verifies the committed seal on each
+// commit message and discards commits with invalid committed seals.
+//
+// Note that when commit messages are discarded they are not removed from the
+// round state database, this should not pose a problem however because if this
+// step is reached again they will again be discarded.
+func (c *core) removeInvalidCommittedSeals() {
+
+	commits := c.current.Commits()
+
+	for _, msg := range commits.Values() {
+		var commit *istanbul.CommittedSubject
+		msg.Decode(&commit)
+		err := c.verifyCommittedSeal(commit, c.current.GetValidatorByAddress(msg.Address))
+		if err != nil {
+			commits.Remove(msg.Address)
```

Comment:
+1 we should definitly do that as we will want to do something in response, possibly Governance slashing.

Reply:
+1

Reply:
So we'd need to extend slashing to be able to do programatic slashing based on bad BLS signatures. I worry that adding the bitmap would then require proof of the bad signature to be tied to the block (but if you don't have the bad bitmap, you could submit the bad signature).

Just logging the address would enable governance slashing / bringing the bad behavior to light.

Reply:
How would logging the address enable governance slashing? How do you verify someone's log?

Reply:
Good question. A single log on its own couldn't really be trusted. I think it could be combined with multiple logs from multiple validators, looking at missed signatures (a validator with a bad signature would not be included). It'd also let us know if we need to modify the client also log out the bad BLS signature to then programmatically slash.

I'm starting to think of logging as an incomplete solution, but without it we don't even know that we have a problem.

Reply:
If possible, we should log here the raw message including signature. With that we would have strong evidence of misbehavior

Reply:
I understand your hesitation about bloating the logs, but don't think it really applies here. We should only log if there is misbehavior, which would be an extreme event which we would want to respond to and address quickly. In this investigation, the signature would be absolutely invaluable, and unless it persisted and retrievable in another way, I feel pretty strongly that we should include it here.

We would expect this this to never happen, so I am personally fine with this solution, having it logged out and taking the manual investigation from there. I don't think this represents a item of work that needs much beyond that as what we really want is a credible threat of response to ward off validators trying to mess with this part of the system. I don't think we need to expand this into a separate item of work.

Reply:
How would we actually respond to and adress this issue? If we saw this in the logs what would we do?

Reply:
If we saw these logs we would want to investigate to determine whether this could have been some kind of bug or malfunction (e.g. wrong signing key being used) or if there is evidence that this is intentional. If intentional, we would publish our findings, using the signature as a core piece of evidence, and potentially follow it up with a Governance slashing proposal.